src/main/java/com/googleintern/wfm/ruleengine/action/RuleValidation.java

```diff
@@ -0,0 +1,214 @@
+        filtersByUserPoolAssignmentBuilder = ImmutableSetMultimap.builder();
+    for (UserPoolAssignmentModel user : existingUserPoolAssignments) {
+      filtersByUserPoolAssignmentBuilder.putAll(user, assignPermissions(user, generatedRules));
+    }
+    return filtersByUserPoolAssignmentBuilder.build();
+  }
+
+  private static ImmutableSet<PoolAssignmentModel> assignPermissions(
+      UserPoolAssignmentModel user, ImmutableSet<RuleModel> generatedRules) {
+    ImmutableMap<ImmutableSet<FilterModel>, ImmutableList<ImmutableSet<Long>>>
+        skillIdsAndRoleIdsByFilters = mapSkillIdsAndRoleIdsByFilters(generatedRules);
+    return generatedRules.stream()
+        .filter(rule -> shouldAssignPermissions(rule, user, skillIdsAndRoleIdsByFilters))
+        .flatMap(
+            rule ->
+                rule.permissionSetIds().stream()
+                    .map(
+                        permissionSetId ->
+                            PoolAssignmentModel.builder()
+                                .setCasePoolId(rule.casePoolId())
+                                .setPermissionSetId(permissionSetId)
+                                .build()))
+        .collect(toImmutableSet());
+  }
+
+  private static ImmutableMap<ImmutableSet<FilterModel>, ImmutableList<ImmutableSet<Long>>>
+      mapSkillIdsAndRoleIdsByFilters(ImmutableSet<RuleModel> generatedRules) {
+    ImmutableMap.Builder<ImmutableSet<FilterModel>, ImmutableList<ImmutableSet<Long>>>
+        skillIdsAndRoleIdsByFiltersBuilder = ImmutableMap.builder();
+    generatedRules.forEach(
+        generatedRule ->
+            generatedRule
+                .filters()
+                .forEach(
+                    orFilters ->
+                        skillIdsAndRoleIdsByFiltersBuilder.put(
+                            orFilters, convertFiltersToSkillIdsAndRoleIds(orFilters))));
+    return skillIdsAndRoleIdsByFiltersBuilder.build();
+  }
+
+  private static boolean shouldAssignPermissions(
+      RuleModel generateRule,
+      UserPoolAssignmentModel user,
+      ImmutableMap<ImmutableSet<FilterModel>, ImmutableList<ImmutableSet<Long>>>
+          skillIdsAndRoleIdsByFilters) {
+    if ((generateRule.workforceId() != user.workforceId())
+        || (generateRule.workgroupId() != user.workgroupId())) {
+      return false;
+    }
+    for (ImmutableSet<FilterModel> orFilters : generateRule.filters()) {
+      if (Sets.intersection(
+                  ImmutableSet.copyOf(user.skillIds()),
+                  skillIdsAndRoleIdsByFilters.get(orFilters).get(0))
+              .isEmpty()
+          && Sets.intersection(
+                  ImmutableSet.copyOf(user.roleSkillIds()),
+                  skillIdsAndRoleIdsByFilters.get(orFilters).get(0))
+              .isEmpty()
+          && Sets.intersection(
+                  ImmutableSet.copyOf(user.roleIds()),
+                  skillIdsAndRoleIdsByFilters.get(orFilters).get(1))
+              .isEmpty()) {
+        return false;
+      }
```

Comment:
From my view this can only tell if user's permissions are partially covered by Rules, or a partial part of the permissions from a Rule is applied to a user. How can we tell if a user is fully covered by Rules and all permissions in the Rule are applied to the user?